[MRG] Reduce codegen cache size

mstimberg opened the pull request:
As discussed in #914, this PR improves Cython's and (to a lesser extent) weave's code generation cache:
1. After successful compilation of a library file for Cython we now delete the corresponding `.cpp` and `.pyx` files, plus any additional temporary library files (I still need to check whether this is working correctly on Windows as well). The library file (e.g. `.so` on Linux) is the only file that we'll ever need later. When I run the Cython test suite with an empty cache, I now get 166MB of cached files instead of 774MB before the change. Quite an improvement! I did not see any noticeable slowdown, and the deletion can be switched off with a preference for debugging (`codegen.runtime.cython.delete_source_files`).
2. When importing Brian 2, if the Cython or weave cache directory take up more than 1GB (`codegen.max_cache_dir_size` preference), the user gets an INFO message like this:
```
INFO       Cache size for target "cython": 1584 MB.
You can call "clear_cache('cython')" to delete all files from the cache or manually
delete files in the "/home/marcel/.cython/brian_extensions" directory. [brian2]
```
3. As the message suggests, `clear_cache('cython')` or `clear_cache('weave')` can be used to delete the respective cache directory.

[Unrelated to the changes here, I had to bump up the numpy dependency to version 1.10, we were actually already using a numpy function in one place that did not exist in numpy 1.9. **EDIT**: I pushed this change directly to master, as tests on all branches were failing because of a related issue]

thesamovar commented on brian2/__init__.py:
```diff
@@ -96,4 +99,51 @@ def _check_dependency_version(name, version):
+        raise ValueError('No cache directory registered for target "%s".' % target)
+    logger.debug('Clearing cache for target "%s" (directory "%s").' %
+                 (target, cache_dir))
+    shutil.rmtree(cache_dir)
```
Are we absolutely 100% sure that ``cache_dir`` could never be something it shouldn't be?! :)

mstimberg replied:
Good point. Maybe we should ask the user for confirmation after displaying the name of the directory? Checking for file types sounds like a good idea as well.

mstimberg commented:
I added a check in `clear_cache` that all the files in the directory have an expected extension as you suggested. It bails out with an error otherwise (`IOError`, not sure that's a great choice but I did not immediately find a better fit). I also added a test that checks that it refuses to delete the directory when a file is present that should not be there. There's no test that clearing the cache actually works because this is not so easy to test: we run tests in parallel and if one test deletes the cache directory then the other tests might not be happy about it...

I think it should be safe enough like this, asking the user for explicit confirmation is maybe too much -- after all, they need to explicitly call `clear_cache`.

thesamovar commented:
Yep - I'm happy with that!
